src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,375 @@
+  {B : set β} {C : set γ} {n : ℕ}
+
+@[to_additive]
+lemma map_prod_eq_map_prod [freiman_hom_class F A β n] (f : F) {s t : multiset α}
+  (hsA : ∀ ⦃x⦄, x ∈ s → x ∈ A) (htA : ∀ ⦃x⦄, x ∈ t → x ∈ A) (hs : s.card = n) (ht : t.card = n)
+  (h : s.prod = t.prod) :
+  (s.map f).prod = (t.map f).prod :=
+freiman_hom_class.map_prod_eq_map_prod' f hsA htA hs ht h
+
+namespace freiman_hom
+
+@[to_additive]
+instance fun_like : fun_like (A →*[n] β) α (λ _, β) :=
+{ coe := to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr' }
+
+@[to_additive]
+instance freiman_hom_class : freiman_hom_class (A →*[n] β) A β n :=
+{ map_prod_eq_map_prod' := map_prod_eq_map_prod' }
+
+/-- Helper instance for when there's too many metavariables to apply
+`fun_like.has_coe_to_fun` directly. -/
+@[to_additive]
+instance : has_coe_to_fun (A →*[n] β) (λ _, α → β) := ⟨to_fun⟩
+
+initialize_simps_projections freiman_hom (to_fun → apply)
+
+@[simp, to_additive]
+lemma to_fun_eq_coe (f : A →*[n] β) : f.to_fun = f := rfl
+
+@[ext, to_additive]
+lemma ext ⦃f g : A →*[n] β⦄ (h : ∀ x, f x = g x) : f = g := fun_like.ext f g h
+
+@[simp, to_additive]
+lemma coe_mk (f : α → β) (h : ∀ s t : multiset α, (∀ ⦃x⦄, x ∈ s → x ∈ A) → (∀ ⦃x⦄, x ∈ t → x ∈ A) →
+  s.card = n → t.card = n → s.prod = t.prod → (s.map f).prod = (t.map f).prod) :
+  ⇑(mk f h) = f := rfl
+
+@[simp, to_additive] lemma mk_coe (f : A →*[n] β) (h) : mk f h = f := ext $ λ _, rfl
+
+/-- The identity map from a commutative monoid to itself. -/
+@[to_additive "The identity map from an additive commutative monoid to itself.", simps]
+protected def id (A : set α) (n : ℕ) : A →*[n] α  :=
+{ to_fun := λ x, x, map_prod_eq_map_prod' := λ s t _ _ _ _ h, by rw [map_id', map_id', h] }
+
+/-- Composition of Freiman homomorphisms as a Freiman homomorphism. -/
+@[to_additive "Composition of additive Freiman homomorphisms as an additive Freiman homomorphism."]
+protected def comp (f : B →*[n] γ) (g : A →*[n] β) (hAB : A.maps_to g B) : A →*[n] γ :=
+{ to_fun := f ∘ g,
+  map_prod_eq_map_prod' := λ s t hsA htA hs ht h, begin
+    rw [←map_map,
+    map_prod_eq_map_prod f _ _ ((s.card_map _).trans hs) ((t.card_map _).trans ht)
+      (map_prod_eq_map_prod g hsA htA hs ht h), map_map],
+    { simpa using (λ a h, hAB (hsA h)) },
+    { simpa using (λ a h, hAB (htA h)) }
+  end }
+
+@[simp, to_additive]
+lemma coe_comp (f : B →*[n] γ) (g : A →*[n] β) {hfg} : ⇑(f.comp g hfg) = f ∘ g := rfl
+
+@[to_additive]
+lemma comp_apply (f : B →*[n] γ) (g : A →*[n] β) {hfg} (x : α) : f.comp g hfg x = f (g x) := rfl
+
+@[to_additive]
+lemma comp_assoc (f : A →*[n] β) (g : B →*[n] γ) (h : C →*[n] δ) {hf hhg hgf}
+  {hh : A.maps_to (g.comp f hgf) C} :
+  (h.comp g hhg).comp f hf = h.comp (g.comp f hgf) hh := rfl
+
+@[to_additive]
+lemma cancel_right {g₁ g₂ : B →*[n] γ} {f : A →*[n] β} (hf : function.surjective f) {hg₁ hg₂} :
+  g₁.comp f hg₁ = g₂.comp f hg₂ ↔ g₁ = g₂ :=
+⟨λ h, ext $ hf.forall.2 $ fun_like.ext_iff.1 h, λ h, h ▸ rfl⟩
+
+-- @[to_additive]
```

Comment:
I presume you want this removed before merge?

Reply:
TODO